File: packages/core/src/sfv/serialize.ts
import type { BareItem, InnerList, Item, Member, Parameters } from './types.js'

/**
 * RFC 9651 §4.1 serialization.
 *
 * Badge only serializes when it is the *signer* (the testkit, the directory
 * helper). Verification never round-trips a received field through here — see
 * the note on `DictionaryEntry.source`.
 */

export class SfSerializeError extends Error {
  override readonly name = 'SfSerializeError'
}

const MAX_INTEGER = 999_999_999_999_999
const MIN_INTEGER = -999_999_999_999_999

export function serializeBareItem(item: BareItem): string {
  switch (item.type) {
    case 'integer':
      if (!Number.isInteger(item.value)) throw new SfSerializeError('integer must be integral')
      if (item.value > MAX_INTEGER || item.value < MIN_INTEGER) {
        throw new SfSerializeError('integer out of range')
      }
      return String(item.value)
    case 'decimal':
      return serializeDecimal(item.value)
    case 'string':
      return serializeString(item.value)
    case 'token':
      if (!/^[a-zA-Z*][!#$%&'*+\-.^_`|~\w:/]*$/.test(item.value)) {
        throw new SfSerializeError(`invalid token: ${item.value}`)
      }
      return item.value
    case 'binary':
      return `:${base64(item.value)}:`
    case 'boolean':
      return item.value ? '?1' : '?0'
    case 'date':
      if (!Number.isInteger(item.value)) throw new SfSerializeError('date must be integral')
      return `@${item.value}`
    case 'displaystring':
      return serializeDisplayString(item.value)
  }
}

function serializeDecimal(value: number): string {
  if (!Number.isFinite(value)) throw new SfSerializeError('decimal must be finite')
  const rounded = Math.round(value * 1000) / 1000
  if (Math.abs(Math.trunc(rounded)).toString().length > 12) {
    throw new SfSerializeError('decimal has too many integer digits')
  }
  const text = rounded.toString()
  return text.includes('.') ? text : `${text}.0`
}

function serializeString(value: string): string {
  let out = '"'
  for (const c of value) {
    const code = c.charCodeAt(0)
    if (code < 0x20 || code > 0x7e) throw new SfSerializeError('string must be printable ASCII')
    if (c === '"' || c === '\\') out += '\\'
    out += c
  }
  return `${out}"`
}

function serializeDisplayString(value: string): string {
  let out = '%"'
  for (const byte of new TextEncoder().encode(value)) {
    if (byte === 0x25 || byte === 0x22 || byte <= 0x1f || byte >= 0x7f) {
      out += `%${byte.toString(16).padStart(2, '0')}`
    } else {
      out += String.fromCharCode(byte)
    }
  }
  return `${out}"`
}

function base64(bytes: Uint8Array): string {
  let bin = ''
  for (const b of bytes) bin += String.fromCharCode(b)
  return btoa(bin)
}

export function serializeParameters(params: Parameters): string {
  let out = ''
  for (const [key, value] of params) {
    out += `;${key}`
    if (!(value.type === 'boolean' && value.value)) out += `=${serializeBareItem(value)}`
  }
  return out
}

export function serializeItem(item: Item): string {
  return serializeBareItem(item.value) + serializeParameters(item.params)
}

export function serializeInnerList(list: InnerList): string {
  const inner = list.items.map(serializeItem).join(' ')
  return `(${inner})${serializeParameters(list.params)}`
}

export function serializeMember(member: Member): string {
  return member.kind === 'item' ? serializeItem(member) : serializeInnerList(member)
}

export function serializeList(members: readonly Member[]): string {
  return members.map(serializeMember).join(', ')
}

export function serializeDictionary(dict: ReadonlyMap<string, Member>): string {
  const parts: string[] = []
  for (const [key, member] of dict) {
    if (member.kind === 'item' && member.value.type === 'boolean' && member.value.value) {
      parts.push(key + serializeParameters(member.params))
    } else {
      parts.push(`${key}=${serializeMember(member)}`)
    }
  }
  return parts.join(', ')
}

// -- convenience constructors, used heavily by the signer and tests ----------

export const sf = {
  integer: (value: number): BareItem => ({ type: 'integer', value }),
  decimal: (value: number): BareItem => ({ type: 'decimal', value }),
  string: (value: string): BareItem => ({ type: 'string', value }),
  token: (value: string): BareItem => ({ type: 'token', value }),
  binary: (value: Uint8Array): BareItem => ({ type: 'binary', value }),
  boolean: (value: boolean): BareItem => ({ type: 'boolean', value }),
  date: (value: number): BareItem => ({ type: 'date', value }),
  displayString: (value: string): BareItem => ({ type: 'displaystring', value }),
  item: (value: BareItem, params: Parameters = new Map()): Item => ({
    kind: 'item',
    value,
    params,
  }),
  innerList: (items: readonly Item[], params: Parameters = new Map()): InnerList => ({
    kind: 'inner-list',
    items,
    params,
  }),
} as const

File: sidecar/sfv/serialize.go
package sfv

import (
	"encoding/base64"
	"errors"
	"fmt"
	"strconv"
	"strings"
)

// ErrSerialize reports a value that cannot be represented as a structured field.
var ErrSerialize = errors.New("cannot serialize structured field value")

const (
	maxInteger = 999999999999999
	minInteger = -999999999999999
)

// SerializeBareItem implements RFC 9651 section 4.1.
//
// Serialization is only used on the signing side. Verification never
// round-trips a received field through here; see DictEntry.Source.
func SerializeBareItem(item BareItem) (string, error) {
	switch item.Kind {
	case KindInteger:
		if item.Int > maxInteger || item.Int < minInteger {
			return "", fmt.Errorf("%w: integer out of range", ErrSerialize)
		}
		return strconv.FormatInt(item.Int, 10), nil
	case KindDecimal:
		rounded := float64(int64(item.Dec*1000+copySign(0.5, item.Dec))) / 1000
		text := strconv.FormatFloat(rounded, 'f', -1, 64)
		if !strings.Contains(text, ".") {
			text += ".0"
		}
		return text, nil
	case KindString:
		var out strings.Builder
		out.WriteByte('"')
		for i := 0; i < len(item.Str); i++ {
			c := item.Str[i]
			if c < 0x20 || c > 0x7e {
				return "", fmt.Errorf("%w: string must be printable ASCII", ErrSerialize)
			}
			if c == '"' || c == '\\' {
				out.WriteByte('\\')
			}
			out.WriteByte(c)
		}
		out.WriteByte('"')
		return out.String(), nil
	case KindToken:
		if !validToken(item.Str) {
			return "", fmt.Errorf("%w: invalid token %q", ErrSerialize, item.Str)
		}
		return item.Str, nil
	case KindBinary:
		return ":" + base64.StdEncoding.EncodeToString(item.Bin) + ":", nil
	case KindBoolean:
		if item.Bool {
			return "?1", nil
		}
		return "?0", nil
	case KindDate:
		return "@" + strconv.FormatInt(item.Int, 10), nil
	case KindDisplayString:
		var out strings.Builder
		out.WriteString(`%"`)
		for _, b := range []byte(item.Str) {
			if b == 0x25 || b == 0x22 || b <= 0x1f || b >= 0x7f {
				out.WriteString(fmt.Sprintf("%%%02x", b))
			} else {
				out.WriteByte(b)
			}
		}
		out.WriteByte('"')
		return out.String(), nil
	}
	return "", fmt.Errorf("%w: unknown kind", ErrSerialize)
}

func copySign(magnitude, sign float64) float64 {
	if sign < 0 {
		return -magnitude
	}
	return magnitude
}

func validToken(s string) bool {
	if s == "" {
		return false
	}
	if !(isAlpha(s[0]) || s[0] == '*') {
		return false
	}
	for i := 1; i < len(s); i++ {
		c := s[i]
		if !(isTchar(c) || c == ':' || c == '/') {
			return false
		}
	}
	return true
}

func SerializeParams(params *Params) (string, error) {
	var out strings.Builder
	for _, key := range params.Keys() {
		value, _ := params.Get(key)
		out.WriteByte(';')
		out.WriteString(key)
		if value.Kind == KindBoolean && value.Bool {
			continue
		}
		text, err := SerializeBareItem(value)
		if err != nil {
			return "", err
		}
		out.WriteByte('=')
		out.WriteString(text)
	}
	return out.String(), nil
}

func SerializeItem(item Item) (string, error) {
	value, err := SerializeBareItem(item.Value)
	if err != nil {
		return "", err
	}
	params, err := SerializeParams(item.Params)
	if err != nil {
		return "", err
	}
	return value + params, nil
}

func SerializeInnerList(list InnerList) (string, error) {
	parts := make([]string, 0, len(list.Items))
	for _, item := range list.Items {
		text, err := SerializeItem(item)
		if err != nil {
			return "", err
		}
		parts = append(parts, text)
	}
	params, err := SerializeParams(list.Params)
	if err != nil {
		return "", err
	}
	return "(" + strings.Join(parts, " ") + ")" + params, nil
}

func SerializeMember(member Member) (string, error) {
	if member.IsInnerList {
		return SerializeInnerList(member.InnerList)
	}
	return SerializeItem(member.Item)
}

func SerializeList(members []Member) (string, error) {
	parts := make([]string, 0, len(members))
	for _, member := range members {
		text, err := SerializeMember(member)
		if err != nil {
			return "", err
		}
		parts = append(parts, text)
	}
	return strings.Join(parts, ", "), nil
}

// SerializeDictionary re-serializes a dictionary from its parsed members,
// eliding "=?1" for boolean-true members as RFC 9651 requires.
func SerializeDictionary(dict *Dictionary) (string, error) {
	parts := make([]string, 0, dict.Len())
	for _, key := range dict.Keys() {
		entry, _ := dict.Get(key)
		member := entry.Member
		if !member.IsInnerList && member.Item.Value.Kind == KindBoolean && member.Item.Value.Bool {
			params, err := SerializeParams(member.Item.Params)
			if err != nil {
				return "", err
			}
			parts = append(parts, key+params)
			continue
		}
		text, err := SerializeMember(member)
		if err != nil {
			return "", err
		}
		parts = append(parts, key+"="+text)
	}
	return strings.Join(parts, ", "), nil
}
